File: 84-LargestRectangleInHistogram/84-LargestRectangleInHistogram.py
# Last updated: 5/29/2026, 11:59:27 AM
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        n=len(heights)
        stack=[]
        max_ele=0
        for i in range(n):
            while stack and heights[i]<heights[stack[-1]]:
                h=heights[stack.pop()]
                if not stack:
                    width=i
                else:
                    width=i-stack[-1]-1
                max_ele=max(max_ele,h*width)
            stack.append(i)
        while stack:
            h=heights[stack.pop()]
            if not stack:
                width=n
            else:
                width=n-stack[-1]-1
            max_ele=max(max_ele,h*width)
        return max_ele

File: 84-LargestRectangleInHistogram/84-LargestRectangleInHistogram.java
// Last updated: 5/29/2026, 11:59:33 AM
class Solution {
    public int largestRectangleArea(int[] heights) {
        Stack<Integer> st = new Stack<>();
        int max = 0;

        for (int i = 0; i < heights.length; i++) {

            while (!st.isEmpty() && heights[i] < heights[st.peek()]) {
                int h = heights[st.pop()];
                int width;

                if (st.isEmpty()) {
                    width = i;
                } else {
                    width = i - st.peek() - 1;
                }

                max = Math.max(max, h * width);
            }

            st.push(i); 
        }
        int n = heights.length;
        while (!st.isEmpty()) {
            int h = heights[st.pop()];
            int width;

            if (st.isEmpty()) {
                width = n;
            } else {
                width = n - st.peek() - 1;
            }

            max = Math.max(max, h * width);
        }

        return max;
    }
}
